packages/lib/src/helpers/encodeCall.ts

```diff
@@ -1,15 +1,97 @@
+/*
+ * Utility that wraps around ethereumjs-abi, providing additional validation
+ * on passed type/value pair parameters.
+ */
+
 import abi from 'ethereumjs-abi';
+import * as util from 'ethereumjs-util';
 import BN from 'bignumber.js';
+import _ from 'lodash';
 
-function formatValue(value: any): string {
-  if (typeof(value) === 'number' || BN.isBigNumber(value)) return value.toString();
-  else if (typeof(value) === 'string' && value.match(/\d+(\.\d+)?e(\+)?\d+/)) return (new BN(value)).toString(10);
-  else return value;
-}
+const ERROR_MESSAGE_PREFIX = 'Encoding error';
+const ERROR_MESSAGE = (type: string, value: any) => `${ERROR_MESSAGE_PREFIX} for type ${type} and value ${value.toString()}`;
 
-export default function encodeCall(name: string, args: string[] = [], rawValues: any[] = []): string {
-  const values: string[] = rawValues.map(formatValue);
-  const methodId: string = abi.methodID(name, args).toString('hex');
-  const params: Buffer = abi.rawEncode(args, values).toString('hex');
+export default function encodeCall(name: string, types: string[] = [], rawValues: any[] = []): string {
```

Comment:
We could definitely split the function to:
1) A simpler encodeCall function that performs point 1, i.e. with no validation.
2) An encodeAndParseCall function that runs the validations.
We would then have to identify the points in the code (outside of this file of course) that involve CLI interaction. The tests would have to be modified too, to account for the two different usages.

However, I'm afraid I don't see the point of not validating encodeCall when it's being used programmatically. If the code is being called via Javascript, the user can screw up as well. In fact, although it is less likely, the user can also screw up when using typescript.

Regarding your comment about using ethereumjs-abi's simpleEncode function, I also don't see the advantage of that. We would have to manually build the strings for type and data, instead of just passing it arrays of types and values.

Reply:
> However, I'm afraid I don't see the point of not validating encodeCall when it's being used programmatically. If the code is being called via Javascript, the user can screw up as well. In fact, although it is less likely, the user can also screw up when using typescript.

The point is providing a consistent API with ABI encode. For instance, if the user supplies `"false"` (the string) for a boolean argument in a regular encodeCall, it gets parsed as true, since a non-empty string is truthy in javascript. With the current behaviour, a user would get different results if initializing a contract manually vs via ZeppelinOS, assuming they are using the programmatic interface.

Reply:
Ok great. I think I understand you fully now :)

I changed the code so that encodeCall now performs no validation whatsoever. It simply forwards the incoming data to ethereumjs-abi. And so, when using encodeCall programmatically, (like in tests for example, or from a program using zos-lib as a dependency), there will be no parsing nor validation on the incoming values.

However, when encode call is used from the zos CLI, which does so via the single entry point of ABIs.ts, instead of calling `encodeCall(method.name, argTypes, args)`, it now does `encodeCall(method.name, argTypes, parseCallValues(argTypes, args))`. Note that parseCallValues is a new function in the encodeCall file, that exclusively does the rawValues to values transformation, with type/value validation.